ref/Microsoft.Build/netstandard/Microsoft.Build.cs

```diff
@@ -1489,6 +1490,19 @@ public partial class ProxyTargets
 }
 namespace Microsoft.Build.FileSystem
 {
+    public delegate bool FindPredicate(ref System.ReadOnlySpan<char> fileName);
+    public delegate TResult FindTransform<TResult>(ref System.ReadOnlySpan<char> fileName);
+    public partial interface IDirectoryCache
+    {
+        bool DirectoryExists(string path);
+        System.Collections.Generic.IEnumerable<TResult> EnumerateDirectories<TResult>(string path, Microsoft.Build.FileSystem.FindPredicate predicate, Microsoft.Build.FileSystem.FindTransform<TResult> transform);
+        System.Collections.Generic.IEnumerable<TResult> EnumerateFiles<TResult>(string path, Microsoft.Build.FileSystem.FindPredicate predicate, Microsoft.Build.FileSystem.FindTransform<TResult> transform);
+        bool FileExists(string path);
+    }
+    public partial interface IDirectoryCacheFactory
+    {
+        Microsoft.Build.FileSystem.IDirectoryCache GetDirectoryCacheForProject(Microsoft.Build.Evaluation.Project project);
```

Comment:
@ladipro : there are some gaps between this contract & CPS design:
 1, CPS cache is directory based, not project based, and not complete. So it is more like IDirectoryCache GetCacheForDirectory(string directoryPath). If we don't have cache for a directory, can we fall back to msbuild ones? (I am not sure how often this will hit SDK folders, or parent folder chains. CPS basically caches the project tree itself, and  some glob cones, but it is not a complete list. Otherwise, we will repeat the same logic your team has built for other folders.

2, one key scenario we use this is the initial evaluation of a project or a new configuration (configurations will share the directory cone most of the time, which we always have after evaluating the first configuration). However, that happens in the construction of the Project, and when we receive the instance of the project back in the middle of the construction, it would be problematic -- we don't know the instance yet, and when we access its internal structure, we don't know what is valid and what is not. It is also a dangerous to leak the instance in the middle of the construction through public contracts.

Reply:
oh, i guess we don't have any cache for other folder in msbuild yet. If that is the case, we can implement one directly based on physical file system, if it is not cached in CPS. That might provide extra benefit for us, so we know the probing path in the entire evaluation, which allows us to validate project evaluation cache.  + @arkalyanms

Reply:
maybe CPS can always provide IDirectoryCache, and use our cached folder if it is possible, otherwise just fall back to the file system. It will allow us to capture all probing path during evaluation, and use it in project evaluation cache in the future (@arkalyanms). In that case, maybe the design is fine, except that we will get Project before the constructor finishes, which is still a design issue.

Reply:
maybe some information similar to what is in the ProjectEvaluationStartedEventArgs, ProjectFile, some kind of context, especially the EvaluationId. See my other comments, it is important for us to know the exact evaluation each request comes from, we have already monitor evaluation start/end, which will allow us to collect what belongs a single evaluation.